File: c/boj/3085.c
#include <stdio.h>

int ft_swap(char* a, char* b)
{
    char temp;

    temp = *a;
    *a = *b;
    *b = temp;
}

int chk(int n, char(* board)[50])
{
    int count = 1; // 같은 문자 count
    int result = 0;

    // 가로 확인
    for (int i = 0; i < n; i++)
    {
        count = 1;
        for (int j = 0; j < n - 1; j++)
        {
            if (board[i][j] == board[i][j + 1])
            {
                count++;
            }
            else count = 1;
            if (result < count) result = count;
        }
    }
    // 세로 확인
    for (int i = 0; i < n; i++)
    {
        count = 1;
        for (int j = 0; j < n - 1; j++)
        {
            if (board[j][i] == board[j + 1][i])
            {
                count++;
            }
            else count = 1;
            if (result < count) result = count;
        }
    }
    return (result);
}

int main(void)
{
    int n;
    char board[50][50];
    int answer = 0;
    int result;
    scanf("%d\n", &n);
    for (int i = 0; i < n; i++)
    {
        scanf("%s", &board[i]);
    }

    // board 입력 확인
    // for (int i = 0; i < n; i++)
    // {
    //     for (int j = 0; j < n; j++)
    //     {
    //         printf("%c : %d %d", board[i][j], i, j);
    //     }
    //     printf("\n");
    // }

    for (int i = 0; i < n; i++)
    {
        result = 0;
        for (int j = 0; j < n - 1; j++)
        {
            // 옆 사탕 바꾸기
            ft_swap(&board[i][j], &board[i][j + 1]);
            result = chk(n, board);
            // 사탕 원위치
            ft_swap(&board[i][j + 1], &board[i][j]);
            if (answer < result)
            {
                answer = result;
            }

            // 위아래 사탕 바꾸기
            ft_swap(&board[j][i], &board[j + 1][i]);
            result = chk(n, board);
            // 사탕 원위치
            ft_swap(&board[j + 1][i], &board[j][i]);
            if (answer < result)
            {
                answer = result;
            }
        }
    }
    printf("%d", answer);
    return (0);
}

File: javascript/boj/3085.js
// 사탕 게임
// https://www.acmicpc.net/problem/3085

let path = "input.txt";
// path = "/dev/stdin"

let input = require("fs").readFileSync(path).toString().trim().split("\n");

const N = Number(input[0]);

const board = [];
for (let i = 1; i <= N; i++) {
  board[i - 1] = input[i].trim().split("");
}

function countRow() {
  let max = 0;
  let count = 0;
  let cur;
  for (let i = 0; i < N; i++) {
    count = 0;
    for (let j = 0; j < N; j++) {
      if (!cur || (cur && cur !== board[i][j])) {
        cur = board[i][j];
        count = 0;
      }
      if (cur === board[i][j]) {
        count++;
      }
      if (count > max) {
        max = count;
      }
    }
  }
  return max;
}

function countCol() {
  let max = 0;
  let count = 0;
  let cur;
  for (let i = 0; i < N; i++) {
    count = 0;
    for (let j = 0; j < N; j++) {
      if (!cur || (cur && cur !== board[j][i])) {
        cur = board[j][i];
        count = 0;
      }
      if (cur === board[j][i]) {
        count++;
      }
      if (count > max) {
        max = count;
      }
    }
  }
  return max;
}

// 행으로 바꾸기
let max = 0;
for (let i = 0; i < N; i++) {
  for (let j = 0; j < N - 1; j++) {
    if (board[i][j] !== board[i][j + 1]) {
      let temp = board[i][j];
      board[i][j] = board[i][j + 1];
      board[i][j + 1] = temp;
      let curRowMax = countRow();
      let curColMax = countCol();
      let curMax = curRowMax >= curColMax ? curRowMax : curColMax;
      if (curMax > max) {
        max = curMax;
      }
      // 되돌려 놓기
      temp = board[i][j + 1];
      board[i][j + 1] = board[i][j];
      board[i][j] = temp;
    }
  }
}

// 열로 바꾸기
for (let i = 0; i < N; i++) {
  for (let j = 0; j < N - 1; j++) {
    if (board[j][i] !== board[j + 1][i]) {
      let temp = board[j][i];
      board[j][i] = board[j + 1][i];
      board[j + 1][i] = temp;
      let curRowMax = countRow();
      let curColMax = countCol();
      let curMax = curRowMax >= curColMax ? curRowMax : curColMax;
      if (curMax > max) {
        max = curMax;
      }
      // 되돌려 놓기
      temp = board[j][i];
      board[j][i] = board[j + 1][i];
      board[j + 1][i] = temp;
    }
  }
}

console.log(max);
